src/tasks/qa.py

```diff
@@ -172,3 +193,191 @@ def get_metrics(self, reset=False):
+
+    def load_data_for_path(self, path, split):
+        """ Load data """
+
+        def split_then_tokenize(sent):
+            """ Tokenize questions while preserving @placeholder token """
+            sent_parts = sent.split("@placeholder")
+            assert len(sent_parts) == 2
+            sent_parts = [
+                process_sentence(self.tokenizer_name, s, self.max_seq_len) for s in sent_parts
+            ]
+            return sent_parts[0][:-1] + ["@placeholder"] + sent_parts[1][1:]
+
+        examples = []
+        data = [json.loads(d) for d in open(path, encoding="utf-8")]
+        for item in data:
+            psg_id = item["idx"]
+            psg = process_sentence(self.tokenizer_name, item["passage"]["text"], self.max_seq_len)
+            ent_idxs = item["passage"]["entities"]
+            ents = [item["passage"]["text"][idx["start"] : idx["end"] + 1] for idx in ent_idxs]
+            qas = item["qas"]
+            for qa in qas:
+                qst = split_then_tokenize(qa["query"])
+                qst_id = qa["idx"]
+                if "answers" in qa:
+                    anss = [a["text"] for a in qa["answers"]]
+                else:
+                    anss = []
+                ex = {
+                    "passage": psg,
+                    "ents": ents,
+                    "query": qst,
+                    "answers": anss,
+                    "psg_id": f"{split}-{psg_id}",
+                    "qst_id": qst_id,
+                }
+                examples.append(ex)
+
+        return examples
+
+    def _load_answers(self) -> None:
+        """ """
+        answers = {}
+        for split, split_path in self.files_by_split.items():
+            if split == "test":
+                continue
+            data = [json.loads(d) for d in open(split_path, encoding="utf-8")]
+            for item in data:
+                psg_id = f"{split}-{item['idx']}"
+                for qa in item["qas"]:
+                    qst_id = qa["idx"]
+                    answers[(psg_id, qst_id)] = [a["text"] for a in qa["answers"]]
+        self._answers = answers
+
+    def get_sentences(self) -> Iterable[Sequence[str]]:
+        """ Yield sentences, used to compute vocabulary. """
+        for split in self.files_by_split:
+            if split.startswith("test"):
+                continue
+            path = self.files_by_split[split]
+            for example in self.load_data_for_path(path, split):
+                yield example["passage"]
+                yield example["query"]
+
+    def process_split(self, split, indexers) -> Iterable[Type[Instance]]:
+        """ Process split text into a list of AllenNLP Instances. """
+        is_using_bert = "bert_wpm_pretokenized" in indexers
+
+        def is_answer(x, ys):
+            """ Given a list of answers, determine if x is an answer """
+            return x in ys
+
+        def insert_ent(ent, template):
+            """ Replace ent into template """
```

Comment:
?

Reply:
ReCoRD's queries are Cloze style, so this inserts a possible answer candidate into the "query".